darts/models/forecasting/forecasting_model.py

```diff
@@ -444,105 +696,202 @@ def historical_forecasts(
-            if train_length and len(train) > train_length:
-                train = train[-train_length:]
-
-            # train_cov = covariates.drop_after(pred_time) if covariates else None
-
-            if (not self._fit_called) or retrain_func(
-                counter=_counter,
-                pred_time=pred_time,
-                train_series=train,
-                past_covariates=past_covariates.drop_after(pred_time)
-                if past_covariates and ("past_covariates" in retrain_func_signature)
-                else None,
-                future_covariates=future_covariates.drop_after(
-                    min(pred_time + series.freq * forecast_horizon, series.end_time())
+            # Determines the time index the historical forecasts will be made on
+            historical_forecasts_time_index = (
+                self._get_historical_forecastable_time_index(
+                    series_,
+                    past_covariates_,
+                    future_covariates_,
+                    (retrain is not False) or (not self._fit_called),
                 )
-                if future_covariates and ("future_covariates" in retrain_func_signature)
-                else None,
-            ):
-                self._fit_wrapper(
-                    series=train,
-                    past_covariates=past_covariates,
-                    future_covariates=future_covariates,
+            )
+
+            # We also need the first value timestamp to be used for prediction or training
+            min_timestamp = (
+                historical_forecasts_time_index[0]
+                - (
+                    self.training_sample_time_index_length
+                    if retrain
+                    else self.predict_sample_time_index_length
+                )
+                * series_.freq
+            )
+
+            if historical_forecasts_time_index is None:
+                raise_log(
+                    ValueError(
+                        "Given the provided model, series and covariates, there is no timestamps "
+                        " where we can make a prediction or train the model. "
+                        "Please check the time indexes of the series and covariates."
+                    ),
+                    logger,
                 )
 
-            forecast = self._predict_wrapper(
-                n=forecast_horizon,
-                series=train,
-                past_covariates=past_covariates,
-                future_covariates=future_covariates,
-                num_samples=num_samples,
+            # prepare the start parameter -> pd.Timestamp
+            if start is not None:
+
+                historical_forecasts_time_index = drop_before_index(
+                    historical_forecasts_time_index,
+                    series_.get_timestamp_at_point(start),
+                )
+
+            else:
+                if (retrain is not False) or (not self._fit_called):
```

Comment:
```suggestion
                if retrain or (not self._fit_called):
```

Reply:
Actually the test is not False is needed, as retrain can be something else than False or True, typically 0. (I had such problems in unit tests)

Reply:
Oh, OK!